src/remote/remote.ts

```diff
@@ -427,36 +283,110 @@ export class Remote {
 				dispose: () => labelFormatterDisposable.dispose(),
 			});
 
-			// If the workspace is not in a running state, try to get it running.
-			if (workspace.latest_build.status !== "running") {
-				const updatedWorkspace = await this.maybeWaitForRunning(
-					workspaceClient,
-					workspace,
-					parts.label,
-					binaryPath,
-					featureSet,
-					firstConnect,
+			// Watch the workspace for changes.
+			const monitor = await WorkspaceMonitor.create(
+				workspace,
+				workspaceClient,
+				this.logger,
+				this.vscodeProposed,
+				this.contextManager,
+			);
+			disposables.push(monitor);
+			disposables.push(
+				monitor.onChange.event((w) => (this.commands.workspace = w)),
+			);
+
+			// Wait for workspace to be running and agent to be ready
+			const stateMachine = new WorkspaceStateMachine(
+				parts,
+				workspaceClient,
+				firstConnect,
+				binaryPath,
+				featureSet,
+				this.logger,
+				this.pathResolver,
+				this.vscodeProposed,
+			);
+			disposables.push(stateMachine);
+
+			try {
+				await this.vscodeProposed.window.withProgress(
+					{
+						location: vscode.ProgressLocation.Notification,
+						cancellable: false,
+						title: "Preparing workspace",
+					},
+					async (progress) => {
+						let inProgress = false;
+						let pendingWorkspace: Workspace | null = null;
+
+						await new Promise<void>((resolve, reject) => {
+							const processWorkspace = async (w: Workspace) => {
+								workspace = w;
```

Comment:
Does the notification let us resolve with any value?  I wonder if we could do `resolve(w)` and then `workspace = await this.vscodeProposed.window.WithProgress`.